src/java.base/share/classes/sun/security/ssl/Hybrid.java

```diff
@@ -0,0 +1,462 @@
+                return new KEM.Encapsulated(
+                        new SecretKeyImpl(left.key(), right.key()),
+                        concat(left.encapsulation(), right.encapsulation()),
+                        null);
+            } else {
+                throw new IllegalArgumentException(
+                        "Invalid range for encapsulation: from = " + from +
+                        " to = " + to + ", expected total secret size = " +
+                        (le.secretSize() + re.secretSize()));
+            }
+        }
+
+        @Override
+        public int engineSecretSize() {
+            if (le != null) {
+                return le.secretSize() + re.secretSize();
+            } else {
+                return ld.secretSize() + rd.secretSize();
+            }
+        }
+
+        @Override
+        public int engineEncapsulationSize() {
+            if (le != null) {
+                return le.encapsulationSize() + re.encapsulationSize();
+            } else {
+                return ld.encapsulationSize() + rd.encapsulationSize();
+            }
+        }
+
+        @Override
+        public SecretKey engineDecapsulate(byte[] encapsulation, int from,
+                int to, String algorithm) throws DecapsulateException {
+            var left = Arrays.copyOf(encapsulation, ld.encapsulationSize());
+            var right = Arrays.copyOfRange(encapsulation,
+                    ld.encapsulationSize(), encapsulation.length);
+            if (from == 0 && ld.secretSize() + rd.secretSize() == to) {
+                return new SecretKeyImpl(ld.decapsulate(left),
+                        rd.decapsulate(right));
+            } else {
+                throw new IllegalArgumentException(
+                        "Invalid range for decapsulation: from = " + from +
+                        " to = " + to + ", expected total secret size = " +
+                        (ld.secretSize() + rd.secretSize()));
+            }
+        }
+    }
+
+    public record SecretKeyImpl(SecretKey k1, SecretKey k2)
+            implements SecretKey {
+        @Override
+        public String getAlgorithm() {
+            return "Generic";
+        }
+
+        @Override
+        public String getFormat() {
+            return null;
+        }
+
+        @Override
+        public byte[] getEncoded() {
+            return null;
+        }
+    }
+
+    /**
+     * Hybrid public key combines two underlying public keys (left and right).
+     * Public keys can be transmitted/encoded because the hybrid protocol
+     * requires the public component to be sent.
+     */
+    public record PublicKeyImpl(String algorithm, PublicKey left,
+            PublicKey right) implements PublicKey {
+        @Override
+        public String getAlgorithm() {
+            return algorithm;
+        }
+
+        // getFormat() returns "RAW" if both keys are X509Key; otherwise null.
+        @Override
+        public String getFormat() {
```

Comment:
My understanding is that this will always return "RAW" even if `left` or `right` come from 3rd-party providers and is not `X509Key`.

Reply:
Fixed.